portal/src/DetailsListWithOrdering.tsx

```diff
@@ -0,0 +1,125 @@
+import React from "react";
+import {
+  DetailsList,
+  IDetailsListProps,
+  IColumn,
+  IconButton,
+} from "@fluentui/react";
+import { Context } from "@oursky/react-messageformat";
+
+import styles from "./DetailsListWithOrdering.module.scss";
+
+interface DetailsListWithOrderingProps extends IDetailsListProps {
+  onSwapClicked: (swapUpward: boolean, index?: number) => void;
+  columns: IColumn[];
+  onRenderItemColumn: (
+    item?: any,
+    index?: number,
+    column?: IColumn
+  ) => React.ReactNode;
+  orderColumnMinWidth?: number;
+  orderColumnMaxWidth?: number;
+}
+
+interface OrderColumnButtonsProps {
+  index?: number;
+  onSwapClicked: (swapUpward: boolean, index?: number) => void;
+}
+
+const OrderColumnButtons: React.FC<OrderColumnButtonsProps> = function OrderColumnButtons(
+  props: OrderColumnButtonsProps
+) {
+  const onUpClicked = React.useCallback(() => {
+    props.onSwapClicked(true, props.index);
+  }, [props]);
+  const onDownClicked = React.useCallback(() => {
+    props.onSwapClicked(false, props.index);
+  }, [props]);
+  return (
+    <div>
+      <IconButton
+        className={styles.orderColumnButton}
+        onClick={onDownClicked}
+        iconProps={{ iconName: "ChevronDown" }}
+      />
+      <IconButton
+        className={styles.orderColumnButton}
+        onClick={onUpClicked}
+        iconProps={{ iconName: "ChevronUp" }}
+      />
+    </div>
+  );
+};
+
+export function useOnSwapClicked<T>(
```

Comment:
I guess we do not need a hook for this. I want to keep things simpler.

In my imagination, the user of `DetailsListWithOrdering` should use it in the same way as `DetailsList`, but needs to provide `onSwap(indexA: number, indexB: number) => void`.

This module can export helper functions to swap two elements in an array **without** mutation.

Below is the pseudo code of my idea

```typescript
import DetailsListWithOrdering, { swap } from "./DetailsListWithOrdering";

function Foobar() {
  const [items, setItems] = useState([]);
  const onSwapClick = useCallback((a: number, b: number) => {
    setItems(swap(items, a, b));
  }, [items]);
  return (
    <DetailsListWithOrdering
      items={items}
      onSwapClick={onSwapClick}
    />
  );
}
```